File: nodejs/backend/src/dto/authDto.ts
import { z } from 'zod';
import { User } from '../model/User';

export const signupSchema = z.object({
  name: z.string().trim().min(1, 'name must not be empty').max(220),
  email: z.string().trim().email('must be a valid email address'),
  phone: z.string().trim().min(1, 'phone must not be empty').max(15),
  password: z.string().min(8, 'password must be at least 8 characters'),
});
export type SignupRequestDto = z.infer<typeof signupSchema>;

export const loginSchema = z.object({
  email: z.string().trim().email('must be a valid email address'),
  password: z.string().min(1, 'password must not be empty'),
});
export type LoginRequestDto = z.infer<typeof loginSchema>;

/**
 * Safe outward-facing view of `User`- no `passwordHash`. Always map
 * through this before sending a user back to a client.
 */
export interface UserResponseDto {
  id: string;
  name: string;
  email: string;
  phone: string;
  createdAt: Date;
}

export function toUserResponseDto(user: User): UserResponseDto {
  return {
    id: user.id,
    name: user.name,
    email: user.email,
    phone: user.phone,
    createdAt: user.createdAt,
  };
}

export interface LoginResponseDto {
  token: string;
  user: UserResponseDto;
}

File: rust/backend/src/dto/auth_dto.rs
use crate::model::User;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Deserialize, Validate)]
pub struct SignupRequest {
    #[validate(length(min = 1, max = 120, message = "name must not be empty"))]
    pub name: String,

    #[validate(email(message = "must be a valid email address"))]
    pub email: String,

    #[validate(length(min = 1, max = 30, message = "phone must not be empty"))]
    pub phone: String,

    #[validate(length(min = 8, message = "password must be at least 8 characters"))]
    pub password: String,
}

#[derive(Debug, Deserialize, Validate)]
pub struct LoginRequest {
    #[validate(email(message = "must be a valid email address"))]
    pub email: String,

    #[validate(length(min = 1, message = "password must not be empty"))]
    pub password: String,
}

/// Safe outward-facing view of `User` — note there is no `password_hash` field.
#[derive(Debug, Clone, Serialize)]
pub struct UserResponse {
    pub id: Uuid,
    pub name: String,
    pub email: String,
    pub phone: String,
    pub created_at: DateTime<Utc>,
}

impl From<User> for UserResponse {
    fn from(u: User) -> Self {
        Self {
            id: u.id,
            name: u.name,
            email: u.email,
            phone: u.phone,
            created_at: u.created_at,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct LoginResponse {
    pub token: String,
    pub user: UserResponse,
}
